server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -693,7 +699,7 @@ private static void parseNonDynamicArray(
             boolean fieldWithFallbackSyntheticSource = mapper instanceof FieldMapper fieldMapper
                 && fieldMapper.syntheticSourceMode() == FieldMapper.SyntheticSourceMode.FALLBACK;
             boolean fieldWithStoredArraySource = mapper instanceof FieldMapper fieldMapper
-                && context.sourceKeepModeFromIndexSettings() == Mapper.SourceKeepMode.ARRAYS;
+                && getSourceKeepMode(context, fieldMapper.sourceKeepMode()) != Mapper.SourceKeepMode.NONE;
```

Comment:
question, should this just be: `getSourceKeepMode(context, fieldMapper.sourceKeepMode()) = ARRAYS;`, since this method is for dealing with arrays?

Reply:
We want to keep array source both for `ARRAYS` and `ALL` values, I think.

Reply:
Right, maybe my assumption (which may be wrong)  is that in this method we only deal with arrays, so I thought just checking for ARRAYS should be sufficient.

Reply:
Based on the [definition](https://github.com/elastic/elasticsearch/blob/f3bc281978c2d529e6579914cded6f033f317a37/server/src/main/java/org/elasticsearch/index/mapper/Mapper.java#L38) of `ALL`, the source should be recorded when parsing an array. Maybe I'm missing something?

Reply:
Right, I understand that arrays are included with ALL. I think this code will work. I just wondered whether `getSourceKeepMode(context, fieldMapper.sourceKeepMode()) = ARRAYS;` would work here too, since only arrays should be handled by the `parseNonDynamicArray(...)` method?

Reply:
The `parseNonDynamicArray` method is called when parsing an array in a doc. In both `ARRAYS` and `ALL` cases, we want to record the source . If we just check for `ARRAYS` here, the array source won't get recorded in the `ALL case - this is similar to the check above for `objectMapper.storeArraySource()` that applies to objects.

Reply:
Ok, I see now. This makes sense now.